trio/_core/_asyncgens.py

```diff
@@ -22,17 +39,21 @@ class AsyncGenerators:
     # asyncgens after the system nursery has been closed, it's a
     # regular set so we don't have to deal with GC firing at
     # unexpected times.
-    alive = attr.ib(factory=weakref.WeakSet)
+    alive: weakref.WeakSet[AsyncGeneratorType[object, NoReturn]] | set[
+        AsyncGeneratorType[object, NoReturn]
+    ] = attr.ib(factory=_WEAK_ASYNC_GEN_SET)
 
     # This collects async generators that get garbage collected during
     # the one-tick window between the system nursery closing and the
     # init task starting end-of-run asyncgen finalization.
-    trailing_needs_finalize = attr.ib(factory=set)
+    trailing_needs_finalize: set[AsyncGeneratorType[object, NoReturn]] = attr.ib(
+        factory=_ASYNC_GEN_SET
+    )
```

Comment:
I swear I remember inference working here. Does it not? Let's make a mypy issue about this.

Reply:
Yea just tested removing the annotation and we get this error:
```console
error: Need type annotation for "trailing_needs_finalize"  [var-annotated]
```

Reply:
Hm. Okay.

Reply:
IIRC you can add a `type=...` parameter to `attr.ib`. It might also be that we're using somewhat outdated `attr` constructions